File: app/src/main/java/com/example/mlallemant/mentalbattle/UI/Menu/Fragment/PlayFragment.java
package com.example.mlallemant.mentalbattle.UI.Menu.Fragment;

import android.content.Intent;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.example.mlallemant.mentalbattle.R;
import com.example.mlallemant.mentalbattle.UI.Game.GameActivity;
import com.example.mlallemant.mentalbattle.UI.Menu.MenuActivity;
import com.example.mlallemant.mentalbattle.Utils.DatabaseManager;
import com.example.mlallemant.mentalbattle.Utils.Game;
import com.example.mlallemant.mentalbattle.Utils.Player;
import com.example.mlallemant.mentalbattle.Utils.SearchGameTask;


/**
 * Created by m.lallemant on 10/11/2017.
 */

public class PlayFragment extends Fragment {

    private final static String TAG = "PlayFragment";

    //UI
    private ImageView iv_back;
    private ImageView iv_play;
    private ProgressBar pg_play;
    private TextView tv_info;

    //Utils
    private boolean isSearchingGame = false;
    private Game currentGame;
    private Player currentPlayer;
    private SearchGameTask searchGameTask;
    private DatabaseManager db;
 
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.menu_play_fragment, container, false);

        MenuActivity menuActivity = (MenuActivity) getActivity();
        currentPlayer = menuActivity.getCurrentPlayer();
        db = DatabaseManager.getInstance();

        initUI(v);
        initListener();

        DisplayMetrics metrics = new DisplayMetrics();
        menuActivity.getWindowManager().getDefaultDisplay().getMetrics(metrics);
        Log.e(TAG, "density : " + metrics.density);
        Log.e(TAG, "densityDpi : " + metrics.densityDpi);
        Log.e(TAG, "scaleDensity : " + metrics.scaledDensity);

        return v;
    }

    @Override
    public void onStop() {
        super.onStop();
        cancelSearch();
    }

    private void cancelSearch() {
        if (searchGameTask != null) {
            searchGameTask.cancel(true);
            searchGameTask = null;
        }
        isSearchingGame = false;
        tv_info.setText("Click on Play to search game !");
        iv_play.setImageDrawable(ContextCompat.getDrawable(getActivity(), R.drawable.ic_select_play));
        pg_play.setVisibility(View.INVISIBLE);
    }


    private void initUI(View v) {
        iv_back = (ImageView) v.findViewById(R.id.select_play_iv_back);
        iv_play = (ImageView) v.findViewById(R.id.select_play_iv_play);
        pg_play = (ProgressBar) v.findViewById(R.id.select_play_pg_play);
        tv_info = (TextView) v.findViewById(R.id.select_play_tv_info);

        tv_info.setText("Click on Play to search game !");
    }

    private void initListener() {

        iv_play.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                if (!isSearchingGame) {
                    launchSearchingGameTask();
                } else {
                    cancelSearch();
                }

            }
        });

        iv_back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                returnSelectorFragment();
            }
        });


    }

    private void returnSelectorFragment() {
        SelectorFragment selectorFragment = new SelectorFragment();
        FragmentManager fm = getFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.replace(R.id.menu_fl_select, selectorFragment);
        ft.commit();
    }

    private void launchSearchingGameTask() {
        isSearchingGame = true;
        tv_info.setText("Searching game ...");
        iv_play.setImageDrawable(ContextCompat.getDrawable(getActivity(), R.drawable.ic_select_cancel));
        pg_play.setVisibility(View.VISIBLE);

        currentGame = null;
        searchGameTask = new SearchGameTask(new SearchGameTask.AsyncResponse() {
            @Override
            public void onFinishTask(Game game) {
                updateUI(game);
            }
        });
        searchGameTask.setParams(currentPlayer, currentGame);
        searchGameTask.execute("");
    }

    private void launchGameActivity(Game game) {
        Intent intent = new Intent(getActivity(), GameActivity.class);
        intent.putExtra("idGame", game.getId());
        intent.putExtra("currentPlayerId", currentPlayer.getId());
        startActivity(intent);
        getActivity().finish();
    }


    private void updateUI(Game game) {
        currentGame = searchGameTask.getCurrentGame();
        currentPlayer = searchGameTask.getCurrentPlayer();

        if (game != null) {
            if ((!game.getPlayer1().getId().equals("")) && (!game.getPlayer2().getId().equals(""))) {

                db.insertInProgressGame(game);
                db.initListenerCurrentGame(game);
                db.deleteAvailableGame(game);

                tv_info.setText("Game found !");
                //Launch game
                launchGameActivity(game);

            } else {
                db.deleteAvailableGame(game);
                db.deletePlayerSearchingPlayer(currentPlayer);
            }
            //pb_btn_play.setVisibility(View.GONE);
        } else {
            launchSearchingGameTask();
        }
    }
}

File: app/src/main/java/com/example/mlallemant/mentalbattle/ui/menu/fragment/PlayFragment.kt
package com.example.mlallemant.mentalbattle.ui.menu.fragment

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.example.mlallemant.mentalbattle.R
import com.example.mlallemant.mentalbattle.databinding.MenuPlayFragmentBinding
import com.example.mlallemant.mentalbattle.ui.extention.setImage
import com.example.mlallemant.mentalbattle.ui.extention.toast
import com.example.mlallemant.mentalbattle.ui.game.GameActivity
import com.example.mlallemant.mentalbattle.ui.game.GameActivity.Companion.BUNDLE_EXTRA_CURRENT_PLAYER_ID
import com.example.mlallemant.mentalbattle.ui.game.GameActivity.Companion.BUNDLE_EXTRA_GAME_ID
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity
import com.example.mlallemant.mentalbattle.utils.DatabaseManager
import com.example.mlallemant.mentalbattle.utils.Game
import com.example.mlallemant.mentalbattle.utils.Game.Companion.generateCalculationList
import com.example.mlallemant.mentalbattle.utils.Player
import com.example.mlallemant.mentalbattle.utils.Utils
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

/**
 * Created by m.lallemant on 10/11/2017.
 */
class PlayFragment : Fragment() {

    private var _binding: MenuPlayFragmentBinding? = null
    private val binding get() = _binding!!

    //Utils
    private var isSearchingGame = false
    private var currentGame: Game? = null
    private var currentPlayer: Player? = null
    private lateinit var db: DatabaseManager

    private val compositeDisposable = CompositeDisposable()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = MenuPlayFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val menuActivity = activity as MenuActivity?
        currentPlayer = menuActivity?.currentPlayer
        db = DatabaseManager.getInstance()
        initUI()
        initListener()
    }

    override fun onStop() {
        super.onStop()
        cancelSearch()
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }

    private fun cancelSearch() {
        compositeDisposable.dispose()
        isSearchingGame = false
        with(binding) {
            selectPlayTvInfo.text = getString(R.string.click_to_play)
            selectPlayIvPlay.setImage(R.drawable.ic_select_play, requireContext())
            selectPlayPgPlay.visibility = View.INVISIBLE
        }
    }

    private fun initUI() {
        binding.selectPlayTvInfo.text = getString(R.string.click_onPlay_to_search)
    }

    private fun initListener() {
        binding.selectPlayIvPlay.setOnClickListener {
            if (!isSearchingGame) {
                launchSearchingGameTask()
            } else {
                cancelSearch()
            }
        }
        binding.selectPlayIvBack.setOnClickListener { returnSelectorFragment() }
    }

    private fun returnSelectorFragment() {
        val selectorFragment = SelectorFragment()
        fragmentManager?.beginTransaction()?.replace(R.id.menu_fl_select, selectorFragment)?.commit()
    }

    private fun launchSearchingGameTask() {
        isSearchingGame = true
        with(binding) {
            selectPlayTvInfo.text = getString(R.string.searching)
            selectPlayIvPlay.setImage(R.drawable.ic_select_cancel, requireContext())
            selectPlayPgPlay.visibility = View.VISIBLE
        }
        currentGame = null
        searchGameTask()
    }

    private fun launchGameActivity(game: Game) {
        val intent = Intent(activity, GameActivity::class.java).apply {
            putExtra(BUNDLE_EXTRA_GAME_ID, game.id)
            putExtra(BUNDLE_EXTRA_CURRENT_PLAYER_ID, currentPlayer?.id)
        }
        startActivity(intent)
        activity?.finish()
    }

    private fun updateUI(game: Game?) {
        currentGame = game
        if (game != null) {
            if (game.player1?.id != "" && game.player2?.id != "") {
                db.insertInProgressGame(game)
                db.initListenerCurrentGame(game)
                db.deleteAvailableGame(game)
                binding.selectPlayTvInfo.text = getString(R.string.game_found)
                //Launch game
                launchGameActivity(game)
            } else {
                db.deleteAvailableGame(game)
                db.deletePlayerSearchingPlayer(currentPlayer)
            }
        } else {
            launchSearchingGameTask()
        }
    }

    private fun searchGameTask() {
        compositeDisposable.add(
            Single.defer { Single.just(searchGame()) }.observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .timeout(Utils.SEARCH_TIME.toLong(), TimeUnit.MILLISECONDS)
                .subscribe(
                    { game -> updateUI(game) },
                    { error -> processError(error) }
                )
        )
    }

    private fun searchGame() : Game {
        var returnGame: Game
        val availableGame = db.findAvailableGame()
        if (availableGame == null) {
            // if no game available, we create one
            Log.e(TAG, "no game available")
            val tmpPlayer = Player("", "", 0, 0, 0, 0)
            val id = UUID.randomUUID().toString()
            val game = Game(id, currentPlayer, tmpPlayer, generateCalculationList())
            db.insertAvailableGame(game)
            returnGame = db.getAvailableGame(game)
        } else {
            // else we insert player1 in game available
            Log.e(TAG, "Game available")
            if (availableGame.player1?.id == "") {
                db.insertPlayer1InAvailableGame(currentPlayer, availableGame)
                Log.e(TAG, "Game Player1 inserted")
            } else {
                db.insertPlayer2InAvailableGame(currentPlayer, availableGame)
                Log.e(TAG, "Game Player2 inserted")
            }
            while (true) {
                returnGame = db.getAvailableGame(availableGame)
                if (returnGame?.player1 != null && returnGame.player2 != null) {
                    if (returnGame.player1?.id != "" && returnGame.player2?.id != "") {
                        break
                    }
                }
            }
        }
        return returnGame
    }

    private fun processError(error: Throwable) {
        if (error is TimeoutException) {
            db.deleteAvailableGame(currentGame)
        } else {
            toast(getString(R.string.serach_game_timeout))
            cancelSearch()
            Log.e(javaClass.simpleName, error.message, error)
        }
    }

    companion object {
        private const val TAG = "searchGameTask"
    }
}